Buffered packet processing

Qgel opened the pull request:
Current blocking processing in OusterDriver::processData() causes packets to be lost if the processing of lidar data takes too much time, see Issue #96.

This PR splits the processing of lidar data into two seperate threads. Previously, the code receiving sensor data from the wire
was blocked by the lidar or imu processors, causing dataloss on slower systems.

* Modifies sensor interface so receiving functions get passed a pointer to a buffer to write their data to
* Implements ringbuffers for lidar and imu data in ouster_driver
* Splits processing into two seperate threads:
  * receiveData receives from the wire and fills ringbuffers
  * processData takes packets from the queue, assembles the lidar
    scan and runs the processors
* Threads are implemented as pure threads instead of ros timers,
  as timing inconsistencies for the receiver thread were also
  leading to data loss

Qgel committed:
Simplify lidar scan assembly code

As double-buffering is used for the scan assemnly (in contrast to the ouster example)
we can simply zero out our buffer for the new frame. We then no longer need to
special-case zeroing-out missing packets.

Qgel committed:
Initialize FullRotationAccumulator with batchReady == false

Qgel committed:
Implement buffered packet processing

This commit splits the processing of lidar data into two seperate
threads. Previously, the code receiving sensor data from the wire
was blocked by the lidar or imu processors, causing dataloss on
slower systems (Issue #96).

* Modifies sensor interface so receiving functions get passed a
  pointer to a buffer to write their data to
* Implements ringbuffers for lidar and imu data in ouster_driver
* Splits processing into two seperate threads:
  * receiveData receives from the wire and fills ringbuffers
  * processData takes packets from the queue, assembles the lidar
    scan and runs the processors
* Threads are implemented as pure threads instead of ros timers,
  as timing inconsistencies for the receiver thread were also
  leading to data loss

SteveMacenski commented on ros2_ouster/src/client/lidar_scan.cpp:
```diff
@@ -82,10 +82,6 @@ ScanBatcher::ScanBatcher(size_t w, const sensor::packet_format & pf)
 
 bool ScanBatcher::operator()(const uint8_t * packet_buf, LidarScan & ls)
 {
-  using row_view_t =
```
Any modifications to client code from ouster_example must first be made there to make sure these style in sync. Once these changes are merged there, I'd be happy to make the appropriate changes here so they stay in line with each other

Qgel replied:
The changes to lidar_scan.cpp probably shouldn't even be in this PR. I just initially thought the issue was that the wrong columns got zeroed-out here and simplified the code until I eliminated that possibility.

The code in ouster_example is already out-of-sync with the code in this repo, as the code in ouster_example does not use a double buffer but instead [caches the last packet](https://github.com/ouster-lidar/ouster_example/blob/13ea8e8b8a4951fb630dbc9108666995c8443bf6/ouster_client/src/lidar_scan.cpp#L305). This means that the modifications I made here cannot be made in the ouster_example repo without also implementing the double buffering there.

I can just revert the commit that touches `lidar_scan.cpp` for this PR. Then, if you want, I can make a separate PR for this change and we can discuss it there further?

Qgel replied:
Just had a look at the history in ouster_example. Looks like they explicitly changed this code in [#3e6af5](https://github.com/ouster-lidar/ouster_example/commit/3e6af59400a764042cce9113cdc4cb76902c620f), so I think it's unlikely we can (or need to) get this change into ouster_example. I'll therefore revert the commit here.

SteveMacenski commented on ros2_ouster/include/ros2_ouster/ouster_driver.hpp:
```diff
@@ -153,8 +154,19 @@ class OusterDriver : public lifecycle_interface::LifecycleInterface
-  uint8_t * _lidar_packet_data;
-  uint8_t * _imu_packet_data;
+  // Ringbuffer for raw received lidar and imu packets
+  std::vector<uint8_t> _lidar_packet_buf;
```
Why not a pointer? You end up just getting `data()` which points to it anyway.

Qgel replied:
Hmm I initially planned to make the size of the buffer a parameter, then realized that probably isn't very useful: If we are filling a 'reasonably sized' buffer, we are probably running so slow that a bigger buffer won't help anyway.

Since the buffer size is static anyway, I could therefore just change it to an std::array? I'm just in general averse to manual allocation where it's avoidable I guess.

SteveMacenski replied:
std array works, we should know the exact size of these based on the metadata of the sensor, we shouldn't have to make any buffer size larger than the actual size used

Qgel replied:
I forgot the ring buffer size isn't actually static, as it depends on `_sensor->getPacketFormat().lidar_packet_size`. So std::array won't work here.

I'd prefer to keep this as an std::vector, as we don't really lose anything, and handling raw pointer just always runs the risk of someone in the future forgetting to delete it etc. But in the end it's your codebase of course. So just let me know and I'll change it to a raw pointer if you prefer.

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -136,6 +136,9 @@ void OusterDriver::onConfigure()
     shared_from_this());
   broadcastStaticTransforms(_sensor->getMetadata());
+
+  _lidar_packet_buf.resize(_sensor->getPacketFormat().lidar_packet_size * _packet_buf_sz);
```
Then just `new` it for the size you like

SteveMacenski replied:
(and delete in end)

SteveMacenski commented:
Did you also try compiling with release flags and still see that it couldn't keep up? That seems odd.

Qgel committed:
Revert "Simplify lidar scan assembly code"

This reverts commit 581b35f14ef098041f03131deb18094ea92b98df.
This modification to the client code is not necessary for the
buffered packet processing, and conflicts with changes since
made in the ouster_examples repository.

Qgel commented:
Yes, I was testing in Release of RelWithDebInfo mode the whole time. 

It's not that the processing can't keep up, otherwise, this PR wouldn't help. We would just slowly fill up our buffer until we had to drop some frames. Rather, the blocking during processing (for me 30 ms with `PCL|IMG|IMU`) in processData() caused packets from the wire to be lost.

Qgel commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -202,44 +221,90 @@ void OusterDriver::broadcastStaticTransforms(
+        if (got_imu) {
+          _imu_packet_tail = (_imu_packet_tail + 1) % _packet_buf_sz;
+          if (_imu_packet_tail == _imu_packet_head)
+            RCLCPP_WARN(this->get_logger(), "IMU buffer overrun!");
```
The warning here can potentially trigger even though no overrrun occured due to a race on `_imu_packet_head` (same for the lidar warning above). 

I could either lock before modifying `_imu_packet_head` in `processData()`, or just remove the warning. Any preference?

bertaveira commented:
I have tested this with a OS1-64 and it completely solved the flickering. To stress test it I made the entire computer struggle to see if this would solve even in the worst case scenarios by running at the same time `stress --cpu 20 --io 4 --vm 4` and there are no section drops at all while the ros2 branch drops like crazy when under stress.

SteveMacenski commented:
Thanks for testing, that's good to know!

Qgel committed:
Lock against race on [lidar|imu]_packet_head

Moves the modification of the head indices to before the mutex
unlock. This is only necessary for the warning about buffer
overruns to work correctly. We don't need to lock against the
actual data in the buffer.

Qgel commented:
I'll be gone for the next few days, so I've just pushed the commit that guards against the race in the head indices. The branch should now be in a good state from my side in case you want to merge while I'm away.

SteveMacenski commented on ros2_ouster/include/ros2_ouster/ouster_driver.hpp:
```diff
@@ -153,8 +154,19 @@ class OusterDriver : public lifecycle_interface::LifecycleInterface
+  std::atomic<int> _lidar_packet_head; std::atomic<int> _lidar_packet_tail;
+  std::vector<uint8_t> _imu_packet_buf;
+  std::atomic<int> _imu_packet_head; std::atomic<int> _imu_packet_tail;
+  int _packet_buf_sz = 1024;
```
Is this a static size from somewhere in the software manual as always true?

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -145,11 +148,14 @@ void OusterDriver::onActivate()
+
+  _processing_active = true;
+  _process_thread = std::thread(std::bind(&OusterDriver::processData, this));
+  _recv_thread = std::thread(std::bind(&OusterDriver::receiveData, this));
```
There's still the main thread of this application which isn't being used, is it not possible for one of these to be on a timer to use it? The 2 threads are both `while true`'s essentially with no sleeping so this will run at full CPU rate, which would have massive CPU utilization. Do you not see that this driver is pulling an unnecessarily large amount of compute resources? It seems like `processData` should be on a timer or event based triggering in the main thread so we don't have a dead thread not doing anything but spinning up a new one. Even if we don't use the main thread, we need something to throttle these while loops so other processes / threads can use the resources.

Qgel replied:
None of the threads will spin unnecessarily in this implementation. 
The receive thread blocks on receiving data from the network interface (the client code `select`s on the fd's with a timeout of 1 second). So when there is no data, the the thread will do no work until the os scheduler wakes it as soon as data is available.
The process thread uses a `condition_variable` to wait for a signal from the receive thread that data is available. This will again cause the os to put the thread to sleep until data is available.
Both these mechanisms also ensure that data is processes as soon as possible so we don't introduce any additional latency.

I tried running the receive thread on a timer, but that did not work well. I also don't think it's the right paradigm for us, because we don't really have a periodic task and we are really just setting a very small timer value so our function essentially always runs anyway. I'm not sure how running one of the threads in the main threads instead would work, as that would block work the rcl needs to do (like service calls etc).

Empirically I see no change in cpu utilization on my box before or after this patch.

SteveMacenski replied:
OK, I haven't run into `condition_variable` before so if you say that it doesn't meaningfully impact CPU performance, I'll take a look again under that understanding. Unfortunately I don't easily have access to a ouster setup anymore so that's partially why I'm so cautious to large changes since I myself am not easily able to test changes to verify functionality.

I looked over the use of condition variable and agree :+1:

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -202,44 +221,98 @@ void OusterDriver::broadcastStaticTransforms(
+      = _data_processors.equal_range(ouster::sensor::client_state::IMU_DATA);
+  auto pf = _sensor->getPacketFormat();
+
+  std::unique_lock<std::mutex> ringbuffer_guard(_ringbuffer_mutex);
```
Overall this seems like overkill in complexity for the driver code and it makes this difficult for at least me to read easily. It seems like the use of some basic data structures like queues could really clean this up a lot and prevent overruns or other similar issues. It comes at the cost of some memory allocation, which I would have liked to have avoid, but the ring buffer seems fragile to me in this situation unless we made it very large that there's no way we could be backed up to that level. 1 full rotation doesn't seem to do so, so at minimum we should push that up. 

But I could definitely see myself approving this method if it were better designed. Much of the ring buffer logic / members / etc should be in a separate class that we use in the driver for managing / grabbing packets. Implementing this inside of the driver itself is really not clean or clear. So I suppose that's my major issue that makes me think to use a queue (that it makes the driver code cleaner and clearer). A separate object for handling this data internally that is then used here would be the compromise to use this method but then still have more readable application code for me to maintain.

So options:
- Abstract out some of this stuff for ring buffer population / management / data getting into a separate class that is used so the driver code looks cleaner / clearer. With this broken out into separate functions, it makes it significantly more readable, and we can add unit tests in the future based on that code
- Use a queue or other similar data structure that would also similarly clean up the application level driver code

Qgel replied:
I'd prefer to stay with a ring buffer instead of a queue, as there really is no need to copy the data twice and I think a ring buffer is pretty standard for this kind of producer/consumer pattern. 

I was thinking of moving the ring buffer code to a separate class when I wrote this, but decided to apply the 'rule of three' since I really only needed two instances. I will move the code to a header though when I get back from vacation in a few days so we don't have the arithmetic directly in the driver code. 

Concerning the size, the buffer just needs to be big enough so it can bridge the time the process thread is working on a frame while new data arrives. The size is currently 1024 packets, which I thought should be larger than 1 frame (as per the comment above the timer rate in the code, I assumed there were 1280 packets / s = 128 packets / frame). 1024 is arbitrary, but seemed like a nice round power of two in the ballpark, with some space to spare for the driver to catch up if the system ever runs into a temporary slowdown. As it is, the buffer is about 2 MB which doesn't seem too much of a footprint. 

I'm not sure what exactly you mean with fragile in this case. If the buffer ever runs over, this would simply mean that all as-of-yet unprocessed frames would be discarded. But if that happens it means the system is too slow to process frames in real time anyway, regardless of buffer size. For a queue, we would also need to check for a max size & discard, or run the risk of OOMing a slow system.

SteveMacenski replied:
OK - a separate header would be great so that the driver code is cleaner to read

Qgel committed:
Move ringbuffer code to it's own header

SteveMacenski commented on ros2_ouster/include/ros2_ouster/ouster_driver.hpp:
```diff
@@ -135,13 +137,14 @@ class OusterDriver : public lifecycle_interface::LifecycleInterface
     const std::shared_ptr<ouster_msgs::srv::GetMetadata::Request> request,
     std::shared_ptr<ouster_msgs::srv::GetMetadata::Response> response);
 
+  void processData();
```
Add doxygen

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -177,10 +184,17 @@ void OusterDriver::onCleanup()
-  _process_timer.reset();
   _tf_b.reset();
 
+  _processing_active = false;
```
Shutdown is after `deactivate` so active was already set as false and the threads were already joined.

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -202,44 +216,83 @@ void OusterDriver::broadcastStaticTransforms(
-    std::pair<DataProcessorMapIt, DataProcessorMapIt> key_its;
+void OusterDriver::processData() {
+  std::pair<DataProcessorMapIt, DataProcessorMapIt> key_lidar_its
+            = _data_processors.equal_range(ouster::sensor::client_state::LIDAR_DATA);
```
Some of the indentation is weird -- please make sure to lint the code with `ament_cpplint` and `ament_uncrustify` (you can ignore any errors in the `client` directory if you did not make change there

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -202,44 +216,83 @@ void OusterDriver::broadcastStaticTransforms(
+          // If the ringbuffer is full, this means the processing thread is running too slow to process all frames
+          // Therefore, we push to it anyway, discarding all (old) data in the buffer and emit a warning.
+          if(_lidar_packet_buf->full())
+              RCLCPP_WARN(this->get_logger(), "Lidar buffer overrun!");
```
Another linting thing - this is 4 spaces

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -202,44 +216,83 @@ void OusterDriver::broadcastStaticTransforms(
+        _process_cond.notify_all();
       }
+
+
```
Remove extra space

SteveMacenski commented on ros2_ouster/include/ros2_ouster/ringbuffer.hpp:
```diff
@@ -0,0 +1,108 @@
+  };
+}
+
+#endif  // ROS2_OUSTER__RINGBUFFER_HPP_
```
missing line at end of file

SteveMacenski commented on ros2_ouster/include/ros2_ouster/ringbuffer.hpp:
```diff
@@ -0,0 +1,108 @@
+   */
+  class RingBuffer
+  {
+  private:
```
Make protected and stick at the end of the class definition (e.g. class variables after all methods)

SteveMacenski reviewed (changes requested):
LGTM, just small nits and happy to merge!

Thanks for the effort, this is a good change. Sorry I was a little curmudgeonly, I guess what I really wanted was just the ring buffer separated out.

Qgel committed:
Fix linting issues

Qgel commented:
Whoops, thought I had already done this. Hope I caught everything.
